concordium-contracts-common/src/types.rs

```diff
@@ -1406,6 +1545,97 @@ pub struct Policy<Attributes> {
+        m.insert(k, v);
+    }
+    m.into_iter().collect()
+}
+
+/// Generate a random `u64` value in the given range by shifting a random `u64`
+/// value using the `%` operator. The reason for doing it this way is that the
+/// range generation method is not exposed by QuickCheck's `Gen`.
+#[cfg(feature = "concordium-quickcheck")]
+fn gen_range_u64(g: &mut Gen, range: core::ops::Range<u64>) -> u64 {
+    let i: u64 = quickcheck::Arbitrary::arbitrary(g);
+    i % (range.end - range.start) + range.start
+}
+
+/// Generate a random `u8` value in the given range by shifting a random `u8`
+/// value using the `%` operator. The reason for doing it this way is that the
+/// range generation method is not exposed by QuickCheck's `Gen`.
+#[cfg(feature = "concordium-quickcheck")]
+fn gen_range_u8(g: &mut Gen, range: core::ops::Range<u8>) -> u8 {
+    let i: u8 = quickcheck::Arbitrary::arbitrary(g);
+    i % (range.end - range.start) + range.start
+}
+
+#[cfg(feature = "concordium-quickcheck")]
+fn valid_owned_policy(op: &OwnedPolicy) -> bool {
+    match op {
+        OwnedPolicy {
+            identity_provider: _,
+            created_at,
+            valid_to,
+            items: _,
+        } => created_at <= valid_to,
+    }
+}
+
+#[cfg(feature = "concordium-quickcheck")]
+impl quickcheck::Arbitrary for OwnedPolicy {
+    fn arbitrary(g: &mut Gen) -> OwnedPolicy {
+        let size: u8 = quickcheck::Arbitrary::arbitrary(g);
+        let created_at: Timestamp = quickcheck::Arbitrary::arbitrary(g);
+        // generate `valid_to` date so it's <= `created_at`
+        let valid_to_millis = gen_range_u64(g, created_at.timestamp_millis()..u64::MAX);
+        OwnedPolicy {
+            identity_provider: quickcheck::Arbitrary::arbitrary(g),
+            created_at,
+            valid_to: Timestamp::from_timestamp_millis(valid_to_millis),
+            items: gen_no_dup_kv_vec(g, size as usize),
+        }
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let identity_provider = self.identity_provider;
+        let created_at = self.created_at.clone();
+        let valid_to = self.valid_to;
+        let items0 = self.items.clone();
+        let iter = identity_provider
+            .shrink()
+            .map(move |ip| {
+                let items = items0.clone();
+                created_at.shrink().into_iter().map(move |ca| {
+                    let items = items.clone();
+                    valid_to.shrink().into_iter().map(move |vt| {
+                        items.shrink().into_iter().map(move |it| OwnedPolicy {
+                            identity_provider: ip,
+                            created_at:        ca,
+                            valid_to:          vt,
+                            items:             it,
+                        })
+                    })
+                })
+            })
+            .flatten()
+            .flatten()
+            .flatten()
+            .filter(valid_owned_policy);
```

Comment:
Why can't we use `flat_map` instead of `map + flatten`?

Reply:
Done.